core/sdk-core/src/main/java/software/amazon/awssdk/core/async/AsyncRequestBody.java

```diff
@@ -143,25 +143,33 @@ static AsyncRequestBody fromString(String string) {
     }
 
     /**
-     * Creates a {@link AsyncRequestBody} from a byte array. The contents of the byte array are copied so modifications to the
-     * original byte array are not reflected in the {@link AsyncRequestBody}.
+     * Creates a {@link AsyncRequestBody} from a byte array.
      *
      * @param bytes The bytes to send to the service.
      * @return AsyncRequestBody instance.
      */
     static AsyncRequestBody fromBytes(byte[] bytes) {
-        return new ByteArrayAsyncRequestBody(bytes, Mimetype.MIMETYPE_OCTET_STREAM);
+        return ByteBuffersAsyncRequestBody.from(bytes);
     }
 
     /**
-     * Creates a {@link AsyncRequestBody} from a {@link ByteBuffer}. Buffer contents are copied so any modifications
-     * made to the original {@link ByteBuffer} are not reflected in the {@link AsyncRequestBody}.
+     * Creates a {@link AsyncRequestBody} from a {@link ByteBuffer}.
      *
      * @param byteBuffer ByteBuffer to send to the service.
      * @return AsyncRequestBody instance.
      */
     static AsyncRequestBody fromByteBuffer(ByteBuffer byteBuffer) {
-        return fromBytes(BinaryUtils.copyAllBytesFrom(byteBuffer));
+        return ByteBuffersAsyncRequestBody.of(null, byteBuffer);
+    }
```

Comment:
Same here. Can we add `fromByteBufferUnsafe` instead of changing the default behavior?

Reply:
yes these can all be renamed.
I thought it might be an issue to change the contracts here, although, I'm not sure if `Unsafe` is the right word in this context, as `DirectByteBuffers` are used with `sun.misc.Unsafe` to allow de-allocation of their off heap memory and it may lead to confusion.
I am finding it difficult to come up with another name though, `fromByteBufferDirect` (same issue), `fromMutableByteBuffer` maybe 🤷

One thing to note is that `AsyncRequestBody#fromFile` doesn't provide the same guarantee regarding concurrent modification.

What are your thoughts on the defensive copying?
personally I would say the developer can apply defensive copies themselves before calling these methods as it's not a complex operation, and I'm wondering if this change could be rolled into a major version bump with other changes such as what I've outlined on https://github.com/aws/aws-sdk-java-v2/issues/3928

Reply:
>  I thought it might be an issue to change the contracts here, although, I'm not sure if Unsafe is the right word in this context,DirectByteBuffers are used with sun.misc.Unsafe to allow de-allocation of their off heap memory and it may lead to confusion.

Yeah, "Unsafe" is meant to illustrate a clear delineation between "the SDK is making no assumptions about your input" and "the SDK is making some critical assumptions about your input". It doesn't have anything to do with `sun.misc.Unsafe`, although I agree that might be one potential confusion. Since we've already committed to the "unsafe" naming elsewhere for this case, though, I'd favor sticking with it.

> One thing to note is that AsyncRequestBody#fromFile doesn't provide the same guarantee regarding concurrent modification.

That's a good point. It is indeed inconsistent. I suppose doing things from disk and doing things from memory were just two different states of mind during our development.

> What are your thoughts on the defensive copying? personally I would say the developer can apply defensive copies themselves before calling these methods as it's not a complex operation, and I'm wondering if this change could be rolled into a major version bump with other changes such as what I've outlined on https://github.com/aws/aws-sdk-java-v2/issues/3928

I think it depends on the audience. If we can assume the developer is thoughtful and experienced, then I agree it's most intuitive to not do a defensive copy. Most people we talk with here on github are the thoughtful and experienced types.

Our assumption about the AWS SDK's customers is that they come from all levels of in-a-hurry and experience. A recently-hired intern, who has never worked with Java, and needs to finish up some work that's leftover from the person that they replaced probably won't be thinking about the mutability of the data they're passing around. They're struggling getting through all of the other AWS complexities that exist. We don't want them to miss their deadlines fighting a data mutability bug and then say "that's dumb, why does Java/AWS/the computer/etc. allow that to happen?" There's already enough places for customer's code to go wrong.

Our solution has been to make the obvious path "safe" and the less-obvious path trade some of that safety for performance. We make them use the method called "unsafe" if they want to prevent the copying.

Reply:
Thank you for the great response, I'll make the changes you suggest as soon as I get a free moment

Reply:
done